### ISSUE TITLE
Ubuntu18.04 net5.0 HttpClient SendAsync ignores timeout (via token), hangs for some time and eventually crash the app

### ISSUE THREAD

**Issue by dvitel** [dotnet/runtime#45010]
### Description

Concrete site with this behavior: https://www.goszakup.gov.kz
Expected behavior (as tested on Windows) - 
```The remote certificate is invalid because of errors in the certificate chain: PartialChain```
or at least timeout

Example:
```c#
        static async Task Main(string[] args)
        {        
            var sw = new Stopwatch();
            sw.Start();
            try {
                using var cl = new HttpClient();
                using var cts = new CancellationTokenSource(7000);
                var res = await cl.GetStringAsync("https://www.goszakup.gov.kz", cts.Token); 
                Console.WriteLine($"[{sw.Elapsed}] Res: {res}");
            } catch (Exception e) {
                sw.Stop(); //Never here
                Console.WriteLine($"[{sw.Elapsed}] Err: {e}.");
            }
        }
```

### Configuration

```
.NET SDK (reflecting any global.json):
 Version:   5.0.100
 Commit:    5044b93829

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  18.04
 OS Platform: Linux
 RID:         ubuntu.18.04-x64
 Base Path:   /usr/share/dotnet/sdk/5.0.100/

Host (useful for support):
  Version: 5.0.0
  Commit:  cf258a14b7

.NET SDKs installed:
  2.1.810 [/usr/share/dotnet/sdk]
  2.2.402 [/usr/share/dotnet/sdk]
  3.0.103 [/usr/share/dotnet/sdk]
  5.0.100 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.22 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.22 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.10 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.22 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.3 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.10 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
```

 OpenSSL 1.1.1  11 Sep 2018

### Other information

Same behavior is for SendAsync or GetStringAsync. 
Played with SocketsHttpHandler Ssl settings - did not help 
On run, app hangs around for 30mins on my machine and then crashes with Aborted (core dumped)
The crash dump contains 10K lines of exception call stack - seems that there is a infinite recursion somewhere 
here is two cycles of loop from exception:
```
   at System.Net.Security.SslStream.AuthenticateAsClient(System.Net.Security.SslClientAuthenticationOptions)
   at System.Net.Http.ConnectHelper+<EstablishSslConnectionAsyncCore>d__4.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.ConnectHelper+<EstablishSslConnectionAsyncCore>d__4, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<EstablishSslConnectionAsyncCore>d__4 ByRef)
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Boolean, System.IO.Stream, System.Net.Security.SslClientAuthenticationOptions, System.Threading.CancellationToken)
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(System.Net.Security.SslClientAuthenticationOptions, System.Net.Http.HttpRequestMessage, Boolean, System.IO.Stream, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool+<ConnectAsync>d__82.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.HttpConnectionPool+<ConnectAsync>d__82, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ConnectAsync>d__82 ByRef)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool+<CreateHttp11ConnectionAsync>d__86.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.HttpConnectionPool+<CreateHttp11ConnectionAsync>d__86, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<CreateHttp11ConnectionAsync>d__86 ByRef)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool+<GetHttpConnectionAsync>d__67.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.HttpConnectionPool+<GetHttpConnectionAsync>d__67, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<GetHttpConnectionAsync>d__67 ByRef)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool.GetConnectionAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool+<SendWithRetryAsync>d__72.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.HttpConnectionPool+<SendWithRetryAsync>d__72, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<SendWithRetryAsync>d__72 ByRef)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(System.Net.Http.HttpRequestMessage, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithProxyAuthAsync(System.Net.Http.HttpRequestMessage, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool.SendAsync(System.Net.Http.HttpRequestMessage, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPoolManager.SendAsyncCore(System.Net.Http.HttpRequestMessage, System.Uri, Boolean, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPoolManager.SendAsync(System.Net.Http.HttpRequestMessage, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionHandler.SendAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.RedirectHandler+<SendAsync>d__4.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.RedirectHandler+<SendAsync>d__4, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<SendAsync>d__4 ByRef)
   at System.Net.Http.RedirectHandler.SendAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpMessageHandlerStage.Send(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)
   at System.Net.Http.SocketsHttpHandler.Send(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)
   at System.Net.Http.HttpMessageInvoker.Send(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)
   at System.Net.Http.HttpClient+<SendAsyncCore>d__85.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.HttpClient+<SendAsyncCore>d__85, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<SendAsyncCore>d__85 ByRef)
   at System.Net.Http.HttpClient.SendAsyncCore(System.Net.Http.HttpRequestMessage, System.Net.Http.HttpCompletionOption, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpClient.Send(System.Net.Http.HttpRequestMessage, System.Net.Http.HttpCompletionOption, System.Threading.CancellationToken)
   at System.Net.Http.HttpClient.Send(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Object[], System.Signature, Boolean, Boolean)
   at System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at Internal.Cryptography.Pal.CertificateAssetDownloader+<>c__DisplayClass5_0.<CreateDownloadBytesFunc>b__0(System.String, System.Threading.CancellationToken)
   at Internal.Cryptography.Pal.CertificateAssetDownloader.DownloadAsset(System.String, System.TimeSpan)
   at Internal.Cryptography.Pal.CertificateAssetDownloader.DownloadCertificate(System.String, System.TimeSpan)
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.FindChainViaAia(System.Collections.Generic.List`1<System.Security.Cryptography.X509Certificates.X509Certificate2> ByRef)
   at Internal.Cryptography.Pal.ChainPal.BuildChain(Boolean, Internal.Cryptography.ICertificatePal, System.Security.Cryptography.X509Certificates.X509Certificate2Collection, System.Security.Cryptography.OidCollection, System.Security.Cryptography.OidCollection, System.Security.Cryptography.X509Certificates.X509RevocationMode, System.Security.Cryptography.X509Certificates.X509RevocationFlag, System.Security.Cryptography.X509Certificates.X509Certificate2Collection, System.Security.Cryptography.X509Certificates.X509ChainTrustMode, System.DateTime, System.TimeSpan, Boolean)
   at System.Security.Cryptography.X509Certificates.X509Chain.Build(System.Security.Cryptography.X509Certificates.X509Certificate2, Boolean)
   at System.Net.Security.CertificateValidation.BuildChainAndVerifyProperties(System.Security.Cryptography.X509Certificates.X509Chain, System.Security.Cryptography.X509Certificates.X509Certificate2, Boolean, System.String)
   at System.Net.Security.SecureChannel.VerifyRemoteCertificate(System.Net.Security.RemoteCertificateValidationCallback, System.Net.Security.ProtocolToken ByRef, System.Net.Security.SslPolicyErrors ByRef, System.Security.Cryptography.X509Certificates.X509ChainStatusFlags ByRef)
   at System.Net.Security.SslStream.CompleteHandshake(System.Net.Security.ProtocolToken ByRef, System.Net.Security.SslPolicyErrors ByRef, System.Security.Cryptography.X509Certificates.X509ChainStatusFlags ByRef)
   at System.Net.Security.SslStream+<ForceAuthenticationAsync>d__171`1[[System.Net.Security.SyncReadWriteAdapter, System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Security.SslStream+<ForceAuthenticationAsync>d__171`1[[System.Net.Security.SyncReadWriteAdapter, System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]], System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ForceAuthenticationAsync>d__171`1<System.Net.Security.SyncReadWriteAdapter> ByRef)
   at System.Net.Security.SslStream.ForceAuthenticationAsync[[System.Net.Security.SyncReadWriteAdapter, System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](System.Net.Security.SyncReadWriteAdapter, Boolean, Byte[], Boolean)
   at System.Net.Security.SslStream.ProcessAuthentication(Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Security.SslStream.AuthenticateAsClient(System.Net.Security.SslClientAuthenticationOptions)
   at System.Net.Http.ConnectHelper+<EstablishSslConnectionAsyncCore>d__4.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.ConnectHelper+<EstablishSslConnectionAsyncCore>d__4, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<EstablishSslConnectionAsyncCore>d__4 ByRef)
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Boolean, System.IO.Stream, System.Net.Security.SslClientAuthenticationOptions, System.Threading.CancellationToken)
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(System.Net.Security.SslClientAuthenticationOptions, System.Net.Http.HttpRequestMessage, Boolean, System.IO.Stream, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool+<ConnectAsync>d__82.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.HttpConnectionPool+<ConnectAsync>d__82, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ConnectAsync>d__82 ByRef)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool+<CreateHttp11ConnectionAsync>d__86.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.HttpConnectionPool+<CreateHttp11ConnectionAsync>d__86, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<CreateHttp11ConnectionAsync>d__86 ByRef)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool+<GetHttpConnectionAsync>d__67.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.HttpConnectionPool+<GetHttpConnectionAsync>d__67, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<GetHttpConnectionAsync>d__67 ByRef)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool.GetConnectionAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool+<SendWithRetryAsync>d__72.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.HttpConnectionPool+<SendWithRetryAsync>d__72, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<SendWithRetryAsync>d__72 ByRef)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(System.Net.Http.HttpRequestMessage, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithProxyAuthAsync(System.Net.Http.HttpRequestMessage, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPool.SendAsync(System.Net.Http.HttpRequestMessage, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPoolManager.SendAsyncCore(System.Net.Http.HttpRequestMessage, System.Uri, Boolean, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionPoolManager.SendAsync(System.Net.Http.HttpRequestMessage, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpConnectionHandler.SendAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.RedirectHandler+<SendAsync>d__4.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.RedirectHandler+<SendAsync>d__4, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<SendAsync>d__4 ByRef)
   at System.Net.Http.RedirectHandler.SendAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpMessageHandlerStage.Send(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)
   at System.Net.Http.SocketsHttpHandler.Send(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)
   at System.Net.Http.HttpMessageInvoker.Send(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)
   at System.Net.Http.HttpClient+<SendAsyncCore>d__85.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.HttpClient+<SendAsyncCore>d__85, System.Net.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<SendAsyncCore>d__85 ByRef)
   at System.Net.Http.HttpClient.SendAsyncCore(System.Net.Http.HttpRequestMessage, System.Net.Http.HttpCompletionOption, Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Http.HttpClient.Send(System.Net.Http.HttpRequestMessage, System.Net.Http.HttpCompletionOption, System.Threading.CancellationToken)
   at System.Net.Http.HttpClient.Send(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Object[], System.Signature, Boolean, Boolean)
   at System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at Internal.Cryptography.Pal.CertificateAssetDownloader+<>c__DisplayClass5_0.<CreateDownloadBytesFunc>b__0(System.String, System.Threading.CancellationToken)
   at Internal.Cryptography.Pal.CertificateAssetDownloader.DownloadAsset(System.String, System.TimeSpan)
   at Internal.Cryptography.Pal.CertificateAssetDownloader.DownloadCertificate(System.String, System.TimeSpan)
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.FindChainViaAia(System.Collections.Generic.List`1<System.Security.Cryptography.X509Certificates.X509Certificate2> ByRef)
   at Internal.Cryptography.Pal.ChainPal.BuildChain(Boolean, Internal.Cryptography.ICertificatePal, System.Security.Cryptography.X509Certificates.X509Certificate2Collection, System.Security.Cryptography.OidCollection, System.Security.Cryptography.OidCollection, System.Security.Cryptography.X509Certificates.X509RevocationMode, System.Security.Cryptography.X509Certificates.X509RevocationFlag, System.Security.Cryptography.X509Certificates.X509Certificate2Collection, System.Security.Cryptography.X509Certificates.X509ChainTrustMode, System.DateTime, System.TimeSpan, Boolean)
   at System.Security.Cryptography.X509Certificates.X509Chain.Build(System.Security.Cryptography.X509Certificates.X509Certificate2, Boolean)
   at System.Net.Security.CertificateValidation.BuildChainAndVerifyProperties(System.Security.Cryptography.X509Certificates.X509Chain, System.Security.Cryptography.X509Certificates.X509Certificate2, Boolean, System.String)
   at System.Net.Security.SecureChannel.VerifyRemoteCertificate(System.Net.Security.RemoteCertificateValidationCallback, System.Net.Security.ProtocolToken ByRef, System.Net.Security.SslPolicyErrors ByRef, System.Security.Cryptography.X509Certificates.X509ChainStatusFlags ByRef)
   at System.Net.Security.SslStream.CompleteHandshake(System.Net.Security.ProtocolToken ByRef, System.Net.Security.SslPolicyErrors ByRef, System.Security.Cryptography.X509Certificates.X509ChainStatusFlags ByRef)
   at System.Net.Security.SslStream+<ForceAuthenticationAsync>d__171`1[[System.Net.Security.SyncReadWriteAdapter, System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Security.SslStream+<ForceAuthenticationAsync>d__171`1[[System.Net.Security.SyncReadWriteAdapter, System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]], System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ForceAuthenticationAsync>d__171`1<System.Net.Security.SyncReadWriteAdapter> ByRef)
   at System.Net.Security.SslStream.ForceAuthenticationAsync[[System.Net.Security.SyncReadWriteAdapter, System.Net.Security, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](System.Net.Security.SyncReadWriteAdapter, Boolean, Byte[], Boolean)
   at System.Net.Security.SslStream.ProcessAuthentication(Boolean, Boolean, System.Threading.CancellationToken)
   at System.Net.Security.SslStream.AuthenticateAsClient(System.Net.Security.SslClientAuthenticationOptions)
```

crash is probably due to StackOverflow. If there is no fix but workaround exists (some global settings (like env) etc) - please share.

**Comment by vcsjones** [comment]
I see. That is... interesting. The `root` certificate is available to download over HTTPS and the root is serving it self. It's rather uncommon for AIA to happen over HTTPS. I don't know how exactly how Windows handles this situation. Failing seems right (especially if that is what Windows is doing.)

The intention in the current AIA implementation is to not use HTTPS at all if it is specified in the certificate as HTTPS:

https://github.com/dotnet/runtime/blob/58667e733437aeed5342f8233a4118a43b7ae319/src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs#L1131-L1132

But that doesn't help the situation where HTTP -> HTTPS redirection is happening. So maybe the better solution is to disable `AllowAutoRedirect`, and follow the redirects ourselves. If we end up getting a redirect to a non-HTTP URL (https://, gopher://, whatever) we just fail there.

**Comment by bartonjs** [comment]
Hm, yeah, we definitely want to block an http -> https redirection.  Following a http -> http redirection seems good, but perhaps with a limit of 2 redirects (3 hops).

And, we probably want to service this when we have a solution.

**Comment by wfurt** [comment]
It is possible this used to work as the links point to http:// but perhaps somebody touched the nginx to do wholesale redirects ;( 

It also seems the X509Chain is based on timeout X509ChainPolicy.UrlRetrievalTimeout but HttpClient is based on Cancellation token so it is hard to stick them together in consistency way. It feels like opportunity for better API...

**Comment by vcsjones** [comment]
> And, we probably want to service this when we have a solution.

I'll give it a shot over the weekend.

**Comment by dvitel** [comment]
@wfurt so can I set up somehow (through public API) X509ChainPolicy.UrlRetrievalTimeout as temp workaround? And does it take into account redirects of cert downloads?

**Comment by wfurt** [comment]
unfortunately no @dvitel. This is hidden down in SslStream and there is no public API to manipulate this. 
You can perhaps try to contact the support@ecc.kz listed on the site. I know this may not work but there was at least one another case when reporting issue lead to fixes.

**Comment by wfurt** [comment]
https://www.ssllabs.com/ssltest/analyze.html?d=www.goszakup.gov.kz

**Comment by vcsjones** [comment]
@wfurt A bit unrelated from the issue, my understanding of the CA is that this is the certificate authority that Kazakhstan has set up to perform mass HTTPS decryption, I don't believe it is expected this certificate chain will work and has been put on a disallow list by many certificate trust stores. https://bugzilla.mozilla.org/show_bug.cgi?id=1232689#c14 provides some additional context.

**Comment by geoffkizer** [comment]
> It also seems the X509Chain is based on timeout X509ChainPolicy.UrlRetrievalTimeout but HttpClient is based on Cancellation token so it is hard to stick them together in consistency way. It feels like opportunity for better API...

Yeah, this seems bad. Is there some reason we don't support CancellationToken in these APIs? Is it just a matter of lack of API surface or is there some underlying issue involved? Which particular API are we talking about here?

**Comment by vcsjones** [comment]
> Is there some reason we don't support CancellationToken in these APIs?

I would presume this is because only on Linux do we actually do AIA fetches using HttpClient. On Windows and MacOS the chain building and fetching process is opaque and handled by the OS. I don’t know what or how a CancellationToken would do on MacOS or Windows.

**Comment by wfurt** [comment]
If there is any chance of asynchronocy, we could at least release control back to caller with partial result.

**Comment by vcsjones** [comment]
> If there is any chance of asynchronocy,

Not sure what you mean, but I wouldn't know of a great way do this operation partially. In the case of macOS, `SecTrustEvaluate` does chain building, revocation checking, AIA fetching, etc. all in one native API call. We couldn't say, do revocation checking, check the cancellation token, and skip AIA fetching if the cancellation token is cancelled. In fact even on macOS, `UrlRetrievalTimeout` does nothing.

Windows' API shape is similar to macOS, with the exception that `UrlRetrievalTimeout` works - but not for AIA fetching, only for fetching CRL / OCSP responses.

**Comment by vcsjones** [comment]
I wrote a failing test for this over at https://github.com/dotnet/runtime/compare/master...vcsjones:45010-no-follow-https.

That leaves me with the question of the best way to fix it. Setting `AllowAutoRedirect` to false and re-implementing the logic ourselves [looks tricky][1], more so because we need to do everything with reflection.

Perhaps something new on `HttpClientHandler` like `Func<Uri, bool> AutoRedirectFilter { get; set; }` that allows a callback to be defined as to whether a particular redirect is followed or not would be useful. I could see other folks wanting this for other reasons, like ensuring redirects are only automatically followed for first party domains.

Alternatively, we could "best attempt" to implement the redirect logic since I think some of it doesn't really apply to AIA, like worrying about a POST -> GET method change. 

[1]: https://github.com/dotnet/runtime/blob/d73c65ed402fad658c39d78773d181c4459b62c8/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs#L28

**Comment by wfurt** [comment]
Is the reflection only to get hands on the handler? 

I'm wondering if we can. add some private.internal flag to control the redirection here:
https://github.com/dotnet/runtime/blob/d73c65ed402fad658c39d78773d181c4459b62c8/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs#L126

instead of duplicating the logic.

**Comment by vcsjones** [comment]
> Is the reflection only to get hands on the handler?

The whole thing is done with reflection due to cyclic dependencies.

Could certainly add something that is internal and set with reflection since there is a bunch of reflection going on anyway.

https://github.com/dotnet/runtime/blob/d73c65ed402fad658c39d78773d181c4459b62c8/src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificateAssetDownloader.cs#L136-L146

**Comment by krwq** [comment]
This bug is related to https://github.com/dotnet/announcements/issues/175
It's already fixed in 2.1/3.1/5.0, the PR above is port against master branch.
